server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java

```diff
@@ -337,17 +343,23 @@ protected ConstraintSecurityHandler configureSpnego(Server server,
       AvaticaServerConfiguration config) {
     final String realm = Objects.requireNonNull(config.getKerberosRealm());
     final String principal = Objects.requireNonNull(config.getKerberosPrincipal());
-
-    // A customization of SpnegoLoginService to explicitly set the server's principal, otherwise
-    // we would have to require a custom file to set the server's principal.
-    PropertyBasedSpnegoLoginService spnegoLoginService =
-        new PropertyBasedSpnegoLoginService(realm, principal);
+    final String principalWithoutRealm = principal.split("@", 2)[0];
+    final String[] principalParts = principalWithoutRealm.split("/", 2);
+    final String serviceName = principalParts[0];
+    final String hostname = principalParts[1];
+    final Path keytabPath = config.getKerberosKeytab();
+
+    ConfigurableSpnegoLoginService spnegoLoginService =
+        new ConfigurableSpnegoLoginService(realm, new TestAuthorizationService());
+    spnegoLoginService.setServiceName(serviceName);
+    spnegoLoginService.setHostName(hostname);
+    spnegoLoginService.setKeyTabPath(keytabPath);
 
     // Roles are "realms" for Kerberos/SPNEGO
     final String[] allowedRealms = getAllowedRealms(realm, config);
 
     return configureCommonAuthentication(Constraint.__SPNEGO_AUTH,
-        allowedRealms, new AvaticaSpnegoAuthenticator(), realm, spnegoLoginService);
+        allowedRealms, new ConfigurableSpnegoAuthenticator(), realm, spnegoLoginService);
```

Comment:
Fun fact, can't move to `ConfigurableSpnegoAuthenticator` without ALSO moving to `ConfigurableSpnegoLoginService`. So these are tied together. I tried to separate out the changes but get a class cast exception.